### DISCUSSION TITLE
pattern and map using a list variable

### DISCUSSION THREAD

**Discussion by franzbischoff** [ropensci/targets#342] [Help]
My problem here is the following:

I create a list with idxs for training and test set:

```R   
  tar_target(
    idxs,
    make_idxs_split(length(file_paths), 0.75, seed = tar_seed())
  )
```
`idxs` has `idxs$training` and `idxs$test`.

The following is not working:

```R
  tar_target(training_set,
    read_ecg(file_paths[idxs$training]),
    pattern = map(idxs$training), // or idxs[["training"]] or idxs["training"]
    format = "rds",
    resources = list(compress = "xz")
  )
```

it fails with `Error: invalid pattern: map(idxs$training)`

Digging, I see that the function `settings_validate_pattern()` in `class_settings.R` is where it stops. This because the symbols `$` or `[[` or `[` are not recognized.

I know that the members of a dynamic list cannot be validated beforehand. So this is a bug, or is there a proper way to do this?

Thank you

**Comment by wlandau** [comment]
`map()` and other patterns only accept target names. so you would be able to supply `map(training_ids)` if `training_ids` were its own target. 

```r
# _targets.R
library(targets)
library(stantargets)
list(
  tar_target(idxs, make_idxs_split(length(file_paths), 0.75, seed = tar_seed())),
  tar_target(id_train, idxs$training),
  tar_target(
    training_set,
    read_ecg(file_paths[id_train]),
    pattern = map(id_train)
  )
)
```

If you think those input files are likely to change, consider [tracking them](https://books.ropensci.org/targets/files.html#external-output-files) so the relevant targets rerun when changes happen. This can be tricky with dynamic branching, but [`tarchetypes::tar_files_input()`](https://docs.ropensci.org/tarchetypes/reference/tar_files_input.html) could help.
